harmony/harmonydb/sql/20240823-ipni.sql

```diff
@@ -0,0 +1,165 @@
+-- Table for storing IPNI ads
+CREATE TABLE ipni (
+    order_number BIGSERIAL PRIMARY KEY, -- Unique increasing order number
+    ad_cid TEXT NOT NULL,
+    context_id BYTEA NOT NULL, -- abi.PieceInfo in Curio
+    -- metadata column in not required as Curio only supports one type of metadata(HTTP)
+    is_rm BOOLEAN NOT NULL,
+
+    previous TEXT, -- previous ad will only be null for first ad in chain
+
+    provider TEXT NOT NULL, -- peerID from libp2p, this is main identifier on IPNI side
+    addresses TEXT NOT NULL, -- HTTP retrieval server addresses
+
+    signature BYTEA NOT NULL,
+    entries TEXT NOT NULL, -- CID of first link in entry chain
+
+    unique (ad_cid)
+);
+
+-- This index will help speed up the lookup of all ads for a specific provider and ensure fast ordering by order_number
+CREATE INDEX ipni_provider_order_number ON ipni(provider, order_number);
+
+-- This index will speed up lookups based on the ad_cid, which is frequently used to identify specific ads
+CREATE UNIQUE INDEX ipni_ad_cid ON ipni(ad_cid);
+
+-- This index will speed up lookups based on the ad_cid, which is frequently used to identify specific ads
+CREATE UNIQUE INDEX ipni_context_id ON ipni(context_id, ad_cid, is_rm);
+
+-- Since the get_ad_chain function relies on both provider and ad_cid to find the order_number, this index will optimize that query:
+CREATE INDEX ipni_provider_ad_cid ON ipni(provider, ad_cid);
+
+
+CREATE TABLE ipni_head (
```

Comment:
TODO (me): Maybe this doesn't need to exist and can just be max(ipni.order_number)